feat: Add lease replacement and testcontainers

brndnmtthws opened the pull request:
Introduces the ability for a client to acquire a lease even if an expired one exists, replacing it after a configurable grace period. This prevents leases from getting stuck indefinitely if the holder crashes without releasing.

- Adds `Lease::lease_v()` accessor.
- Replaces manual Docker setup script with `testcontainers` for managing DynamoDB Local in integration tests.
- Updates tests to use `testcontainers` and adds new tests for the replacement logic.
- Separates `ClientBuilder::build()` from `build_and_check_db()`.

brndnmtthws committed:
feat: Add lease replacement and testcontainers

Introduces the ability for a client to acquire a lease even if an
expired one exists, replacing it after a configurable grace period. This
prevents leases from getting stuck indefinitely if the holder crashes
without releasing.

- Adds `ClientBuilder::grace_period()` (default 60s).
- Adds `Client::acquire_or_replace_expired_lease()` which attempts to
  delete a lease if it's expired beyond TTL + grace_period before
  acquiring. Uses conditional delete on `lease_version`.
- Adds backward compatibility for replacing leases without a version
  field.
- Adds `Lease::lease_v()` accessor.
- Replaces manual Docker setup script with `testcontainers` for managing
  DynamoDB Local in integration tests.
- Updates tests to use `testcontainers` and adds new tests for the
  replacement logic.
- Separates `ClientBuilder::build()` from `build_and_check_db()`.

brndnmtthws committed:
cargo fmt

brndnmtthws committed:
don't need container in CI

alexheretic commented:
I don't think we need an extra "grace" concept. The lease being expired should mean the same thing as if it were deleted.

This is discussed in #3. If a client process is killed while holding a lease, currently we simply rely on dynamodb ttl management to delete expired items, so such an expired lease can stick around until the ttl deletion actually happens. This can be improved by simply overwriting expired leases on acquisition, but it is already not possible for a lease to be stuck indefinitely.

brndnmtthws commented:
The main reason for having some grace period is that there can be some slight clock skew, as you cannot guarantee clocks are perfectly in sync, but +/- 60s is probably good enough (assuming you have NTP running properly).

[The DynamoDB docs specify that](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html):

> DynamoDB automatically deletes expired items **within a few days** of their expiration time

"within a few days" is not good enough.

alexheretic commented:
I agree I'd like to see a improvement to the crash scenario, so it doesn't wholly rely on db ttl deletion.

A simple clock skew mitigation is [already documented](https://github.com/alexheretic/dynamodb-lease/blob/main/DESIGN.md#clock-skew). Basically the `lease_ttl_seconds` / expiry is already a configurable grace period for this purpose, I don't see why we'd benefit from an addition one.

That's why I think the improvement should be just overwriting expired leases, since once a lease has expired... it has expired. The db's ttl process is free to delete it and so is anything else.

brndnmtthws commented:
> I agree I'd like to see a improvement to the crash scenario, so it doesn't wholly rely on db ttl deletion.
> 
> A simple clock skew mitigation is [already documented](https://github.com/alexheretic/dynamodb-lease/blob/main/DESIGN.md#clock-skew). Basically the `lease_ttl_seconds` / expiry is already a configurable grace period for this purpose, I don't see why we'd benefit from an addition one.
> 
> That's why I think the improvement should be just overwriting expired leases, since once a lease has expired... it has expired. The db's ttl process is free to delete it and so is anything else.

Okay, makes sense. Do you think we should have separate methods for this, or just add the functionality to the existing API (i.e., do it in `acquire()`).

alexheretic commented:
Yeah I think we can enhance the current acquisition to handle an expired existing lease, so have minimal or ideally no changes to the api.

brndnmtthws committed:
feat: Better handling, testing of expired/released leases

- Add test `try_acquire_replaces_expired` to verify that acquiring a
  lease successfully replaces an existing, expired lease item in
  DynamoDB.
- Update test `release_try_acquire` to include a `GetItem` check,
  ensuring the lease item is deleted from DynamoDB after `release()` is
  called.
- Update documentation for `try_acquire`, `acquire`, and
  `acquire_timeout` to explicitly state that expired leases are replaced
  immediately upon acquisition attempt.

brndnmtthws commented:
Updated the PR based on the feedback, and now it will replace an expired lease with a put.

brndnmtthws committed:
remove redundant sentence

alexheretic commented on tests/usage.rs:
```diff
@@ -293,12 +300,13 @@ async fn init_should_check_table_exists() {
         "{}",
         err
     );
+    let _ = instance.stop().await;
 }
 
 #[tokio::test]
 async fn init_should_check_hash_key() {
     let table_name = "table-with-wrong-key";
-    let db_client = localhost_dynamodb().await;
+    let (db_client, instance) = get_test_db().await;
```
I'm not convinced this is better. We don't need isolated dbs for each test, it's heavier and slower. Also using new dev-dependencies increases build time for this. I'm not really seeing the win?

Maybe leave this for another pr explaining the benefit

alexheretic commented on src/builder.rs:
```diff
@@ -85,33 +88,53 @@ impl ClientBuilder {
+    /// slightly delayed.
+    ///
+    /// Default `60s`.
+    pub fn grace_period(mut self, grace_period: Duration) -> Self {
```
Didn't we discuss this already and there was no need for an additional duration config?

alexheretic commented:
My take from the discussion was that we'd fix the case an expired lease existed and hadn't been reaped yet by overwriting it. This doesn't seem to be that though, I'm not sure what this is though.

alexheretic commented:
I just implemented it myself in #9. Sorry, I have quite limited time and in this case this was more efficient than extending this review.
